# Introduction to Coding (in Python)
## Lesson 5 - NumPy

### NumPy


NumPy is a python library. Libraries are what make python a powerful coding language used by scientists across the world. Your Python implementation will have several built-in libraries. It is possible that NumPy is not built in. Here are some quick instructions for making sure you have the NumPy library

>**Getting Numpy** \
*In Anaconda* \
Open Anaconda Navigator. Use the left hand side of Navigator to get to Environments. Choose your default environment, and search your libraries for Numpy. If it is not installed, install it! You can do this by looking for all libraries, checking the box next to numpy, and hitting install.  \
\
*In terminal* \
If you have a base Python version available via a terminal you can use the command \
`python3 -m pip install numpy` or `python -m pip install numpy`\
to install the numpy library. \
\
For both these installations, you'll probably have to restart this notebook before you import the library here. \
**While you install numpy, you should also get `matplotlib`, as we'll use it in the next lesson.**  \
Other good libraries are `astropy`, `scipy`, and `glob`.

To use a library in Python code, we import it:

In [42]:
import numpy

If the above does not work, make sure the numpy library is installed, then restart this notebook. If that does not work, then restart Jupyter notebook entirely. 

Because some libraries have long names, the import statement can be combined with `as` to rename the library in the code:

In [3]:
import numpy as np

Now we can get to work! 

### NumPy Arrays

Numpy arrays are like lists, but better! They are more intuitive and easier to visualize. 

One big difference between lists and arrays is that Numpy arrays work better when you do not append new data to them. If you know how many entries you need in the array, go ahead and define it with a function like `np.zeros()`, then fill in the entries. 

**Example 1**

Make a 3 by 3 array with $1$ along the diagonal. Below, I print out the array after `np.zeros()`, and after I have edited the diagonal.

In [4]:
myArr = np.zeros([3,3])
print(myArr)
myArr[0,0] = 1
myArr[1,1] = 1
myArr[2,2] = 1
print(myArr)

[[0. 0. 0.]
 [0. 0. 0.]
 [0. 0. 0.]]
[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


The above can be simplified with a different numpy function:

In [8]:
myArr = np.diag([1.0,1.0,1.0])
print(myArr)
myArr = np.diag(np.ones(3))
print(myArr)


[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]
[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


The above code shows two new functions `np.diag` and `np.ones`. Ones gives you an array of 1's instead of an array of zeros. However, ones and zeros work the same - you just give them the dimensions of your array. Diag takes a given 1 dimensional array and puts it along the diagonal of an array of zeros. 

Numpy arrays can get pretty complex. For example, what about an arbitrarily large number?

**Example 2**

In [9]:
n=100
myArr = np.diag(np.ones(n))
print(myArr)

[[1. 0. 0. ... 0. 0. 0.]
 [0. 1. 0. ... 0. 0. 0.]
 [0. 0. 1. ... 0. 0. 0.]
 ...
 [0. 0. 0. ... 1. 0. 0.]
 [0. 0. 0. ... 0. 1. 0.]
 [0. 0. 0. ... 0. 0. 1.]]


Because the above array is really long, python uses ellipses (...) to show less of the array. 

We can also have arrays with more than 2 dimensions:

In [20]:
n=3
myArr = np.zeros([n,n,n])
print(myArr)

[[[0. 0. 0.]
  [0. 0. 0.]
  [0. 0. 0.]]

 [[0. 0. 0.]
  [0. 0. 0.]
  [0. 0. 0.]]

 [[0. 0. 0.]
  [0. 0. 0.]
  [0. 0. 0.]]]


### Grids of numbers... What's the point?

Linear algebra is a basic mathematical formalism that underlies much of physics and computer science (and geology!). Arrays allow us to do linear algebra without writing out a bunch of computations! 

A simple math problem is a *system of equations*:
$$3y + 2x  = 0$$
$$x -6y = 6$$

Most people who have taken algebra can solve this system. However, what if we have 3 or more variables? For example:

$$4x +2y = 0$$
$$3x -8y +3z= 6$$
$$6z -10y= 3$$

Once again, with brute force and time on your hands, most people can figure this out. But the computer can do it MUCH faster:

**Example 3**

In [61]:
A = np.array([[2,3],[1,-6]])
b = np.array([0,6])
solution = np.dot(np.linalg.inv(A),b)
print("For the first system, x = "+format(solution[0],'1.1f')+" and y = "+format(solution[1],'1.1f'))

For the first system, x = 1.2 and y = -0.8


Try the above numbers for yourself, and you'll find they are the correct solution for

$$3y + 2x  = 0$$
$$x -6y = 6$$

Before we go on, I want to explain some linear algebra. The code above results from rearranging the equations into a linear algebra equation:
$$\left[ \begin{array}{cc}
 2 & 3 \\
 1 & -6 
\end{array} \right]
\left[ \begin{array}{c}
 x \\
 y 
\end{array} \right]
= \left[ \begin{array}{c}
 0 \\
 6 
\end{array} \right]$$

The above equation is equivalent to the the system of equations we had before! To multiply in linear algebra, I encourage you to take a look at https://www.mathsisfun.com/algebra/matrix-multiplying.html. 

Now, the above equation can be abbreviated to be 
$$
A \cdot \vec{u} = \vec{b}
$$
where each term is 
$$A = \left[ \begin{array}{cc}
 2 & 3 \\
 1 & -6 
\end{array} \right]$$

$$\vec{u} = \left[ \begin{array}{c}
 x \\
 y 
\end{array} \right]$$

$$\vec{b} = \left[ \begin{array}{c}
 0 \\
 6 
\end{array} \right]$$

Then to solve the equation, we *divide* by $A$. But it is not obvious how to divide by an array. 

Instead, we use something called the Inverse of an array: $A^{-1}$.

Now, there are many formulas and ways to derive the inverse of a matrix. Some matrices do not have an inverse! 

But that's a lot of math. Instead, let's just use the inverse to solve our problem:

$$
\vec{u} = A^{-1} \cdot \vec{b}
$$

And that is what I did in code in Example 3! I multiplied the inverse of $A$ with $\vec{b}$. That is then my solution array $\vec{u}$.

Now do the same thing with the other system of equations: 
> **Problem 1**\
Solve the system of equations below:\
$$4x +2y = 0$$\
$$3x -8y +3z= 6$$\
$$6z -10y= 3$$\
HINT: The above can also be written as \
$$\left[ \begin{array}{ccc}
 4 & 2 & 0 \\
 3 & -8 & 3 \\
 0 & -10 & 6 
\end{array} \right]
\left[ \begin{array}{c}
 x \\
 y \\
 z
\end{array} \right]
= \left[ \begin{array}{c}
 0 \\
 6 \\
 3
\end{array} \right]$$

Hopefully you have started to see the importance of linear algebra and arrays in solving science problems!

### Other parts of NumPy

NumPy does more than just linear algebra and data storage. It also contains many of the various math functions you may have come across. Trigonometric functions, square root, and even the `sinc` function,
$$\mathrm{sinc}(x) = \frac{\sin(\pi x)}{\pi x}$$
are included in NumPy. 

**Example 4**

In [60]:
print(np.pi)
print(np.sin(np.pi/2.0))


3.141592653589793
1.0


It is important to note that `np.sin(angle)` assumes the angle is measured in radians. 

NumPy also has some other functions which contruct lists.

**Example 5**

In [74]:
myLst = np.linspace(0,10,11)
print(myLst)
myLst = np.logspace(0,10,11)
print(myLst)

[ 0.  1.  2.  3.  4.  5.  6.  7.  8.  9. 10.]
[1.e+00 1.e+01 1.e+02 1.e+03 1.e+04 1.e+05 1.e+06 1.e+07 1.e+08 1.e+09
 1.e+10]


Mess around with the above list constructors. The final argument is the number of entries in the list. The first two arguments are the first and last entry. For `linspace`, the list is filled in by adding the step size
$$\Delta x = \frac{\mathrm{End} - \mathrm{Start}}{\mathrm{Num. \, of \, Entries}}$$
to each successive entry. 

`logspace` is similar, but the quantities are spaced logarithmically. This is useful for plotting exponential and logarithmic functions (see next lesson on MatPlotLib).

### Loops + Arrays

Combining loops and arrays can help us solve very difficult problems. First, let's load in some data. NumPy can load text files, like comma separated tables (.csv files):

**Example 6**
In the file `myData.csv,` the grades of 250 students in a Calculus 1 class are recorded. There were 3 tests throughout the semester, and the final column has their grades on the final exam. You can load in the data with the function `np.loadtxt`:

In [39]:
myData = np.loadtxt("myData.csv",delimiter=',',skiprows=1)
print(myData.shape)
print(myData)

(250, 4)
[[ 55.  80.  88.  70.]
 [ 55.  67.  72.  84.]
 [ 56.  64.  77.  65.]
 [ 57.  56.  92.  67.]
 [ 57.  71.  81.  74.]
 [ 57.  71.  56.  89.]
 [ 58.  78.  84.  85.]
 [ 59.  72.  80.  58.]
 [ 60.  79.  59.  77.]
 [ 62.  71.  55.  78.]
 [ 62.  86.  65.  77.]
 [ 63.  83.  74.  81.]
 [ 64.  75.  68.  63.]
 [ 64.  74.  66.  82.]
 [ 65.  55.  63.  81.]
 [ 65.  64.  80.  59.]
 [ 65.  52.  89.  55.]
 [ 66.  80.  69.  88.]
 [ 66.  79.  64.  62.]
 [ 66.  64.  91.  89.]
 [ 66.  56.  62.  82.]
 [ 67.  73.  71.  87.]
 [ 67.  65.  76.  71.]
 [ 67.  55.  78.  75.]
 [ 67.  77.  95.  66.]
 [ 67.  58.  61.  67.]
 [ 67.  73.  79.  77.]
 [ 68.  63.  84.  68.]
 [ 68.  79.  86.  88.]
 [ 68.  74.  82.  82.]
 [ 68.  60.  77.  53.]
 [ 68.  58.  54.  79.]
 [ 69.  67.  85.  69.]
 [ 69.  77.  80.  69.]
 [ 69.  65.  60.  73.]
 [ 69.  78.  69.  88.]
 [ 70.  65.  73.  89.]
 [ 70.  70.  87.  76.]
 [ 71.  85.  91.  81.]
 [ 71.  65.  77.  88.]
 [ 71.  59.  57.  93.]
 [ 71.  60.  70.  78.]
 [ 71.  60.  53.  68.]
 [

Note the extra arguments. There is a delimiter, the character separating each value. Also, if you look at the file, it has headings, which are not numbers. Therefore, we tell NumPy to skip the 1st row, because it does not have data in it. 

We know how indexing works for list. They work similarly for arrays. We use `Lst[i]` to select entry number `i` in the list. 

Notice that `myData` has a shape `(250,4)`. This means the first dimension is each student. So the first student's grades are found with the command `myData[0]`. What if we want the grades for the first test, but all students? We can use a `:` for the index: `myData[:,0]`. This tells NumPy to return all the grades in the 1st column of the array.

>**Problem 2**\
Find the mean score for the first test, using `np.average(data)`, where `data` is a just the grades from the first test.

Now, assuming the data on each test is [normally distributed](https://en.wikipedia.org/wiki/Normal_distribution), we can also calculate the standard deviation of the final test with:

In [43]:
print(np.std(myData[:,3]))

10.820930828722638


That means about $68\%$ of students fell within 10.8 points of the average score. The percentage is a result of our assumption that the scores follow a normal distribution.

>**Problem 3** \
Compare the average and standard deviation of each test with the average and standard deviation of the final grades. \
    Hint: `np.average(myData,axis=1)` will return a list of the final grades. The `axis` parameter tells whether to average over columns or rows. `axis=0` is the default, and averages each column of the array. 